### DISCUSSION TITLE
Classification store attributes inheritance

### DISCUSSION THREAD

**Discussion by tmeinsier** [pimcore/pimcore#8113] [Q&A / Support]
Hello,

In Data object, i've a tree with folder, and product and reference class.
I want to associate a collection store for a folder, and display fields of this collection store on product form, or reference form to complete them.

For example, i've a folder named "Cars" and i associate collection store "cars" (wich contain some group of fields for car object).
When i create a product (or reference), i want to see all fields of "cars" collection store, and i can fill values.

How is it possible ?
Can i do this on association classication on folder, or folder doesn't permit this, and i must create a specific class object and use it for my tree ?

Thank you.

**Comment by butlersupply** [comment]
Inheritance only works for objects of the same class.  So for this to work, you would need to use "product" data object for your tree and assign your classification store to these.  Then when you add products under these, they will inherit the classification store from their parent.  

I didn't like this.  So what i did on my last project was to create a "folder" class.  I assigned my classification store to it, and then on my product class I have have another classification store that uses calculated fields to "inherit" attribute names from the parent, but still let me input attribute values for those names at the product level.  This allows me to have generic classification stores.  Where the parent store holds the attribute 'name' and the child store get's it's name from the parent but it's value from the child.  And the classes can then be different.  This worked out great for my use case.

Another thing that I've done to work around this is to create different views of the data class, in the class definition.  So when you use a product data class as a 'folder'.  The user sees the 'folder' view when opened.  And when you use a product data class as a 'product', the user sees the 'product' view when opened.  Inheritance works because everything is the same class.  You can also make different icons based on how the product is used (folder or product or reference, etc)

**Reply by tmeinsier** [reply]
Thank you for your answer.
Before posting here, i've searched and tried your solution with object class folder to associate classification store to it, but i didn't seen how to inherit this on product below !
But, in your project, when you use calculated fields in classification store, you use calculated values on attribute ?
So, in this case, if i've associated a classification group on folder class above product, when i add product in this folder, how is it possible to find the same classifcation group (with same classification key attributes) on the product ? (the inheritance group classification doesn't exist)
You use/develop bundle for this ?
I would understand how is it work, and know what is it possible or not without bundle.
Thank you again.

**Reply by butlersupply** [reply]
At the folder level I use two classification stores.  The first holds the attribute names.  It looks like this:

Group Collection - Attribute Name Collection
 + Group - Attribute Group 1
 ++ Key Definitions - name, type=input
 + Group - Attribute Group 2
 ++ Key Definitions - name, type=input
 + Group - Attribute Group 3
 ++ Key Definitions - name, type=input
...

So in this, there is only one key called 'name'.  I have 50 "Groups" called "Attribute Group 1" through "Attribute Group 50"

The second class store at the folder level holds folder level options.  It looks like this:

Group Collection - Attribute Options Collection
 + Group - Attribute Group 1 
 ++ Key Definitions - name, type=Calculated Value
 ++ Key Definitions - attributeType, type=input
 ++ Key Definitions - displaySequence, type=number
 + Group - Attribute Group 2
 ++ Key Definitions - name, type=Calculated Value
 ++ Key Definitions - attributeType, type=input
 ++ Key Definitions - displaySequence, type=number
 + Group - Attribute Group 3
...

There are 50 of these.  Note the group names match.  Also note that "name" is a calculated field.  it gets it's value from the Attribute Name Collection.

At the product level I have one more field collection.  It looks like this:

Group Collection - Attribute Value Collection
 + Group - Attribute Group 1 
 ++ Key Definitions - name, type=Calculated Value
 ++ Key Definitions - uom, type=input
 ++ Key Definitions - value, type=input

Again there are 50 groups with group names that match.  The name is a calculated field that matches the Attribute Name Collection for that group.

So in my case, I only 'inherit" the attribute names from the "folder" data object class.  But in your design you can carry forward values using calculated fields in the product's classifications store.

Note that you can't override calculated fields.  So if you have a product that inherits from another, and you want to override a value in only the child product - you cannot.  A work-around would be to have an "input" field key in the product class store that, if populated, overrides the calculated field.

I also have listener code that adds the classification store to the product automatically.  

As a side note, this approach was easier than creating hundreds of classification stores for each of my attribute sets.  YMMV.
